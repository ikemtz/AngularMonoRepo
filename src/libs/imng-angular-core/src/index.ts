/* eslint-disable-next-line @typescript-eslint/no-empty-function */
export function main(): void {} //NOSONAR - This is strictly for the build process
