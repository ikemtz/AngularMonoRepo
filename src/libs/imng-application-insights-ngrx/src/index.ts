export * from './lib/imai-application-insights-ngrx.module';
export * from './lib/app-insights-monitoring.service';
export * from './lib/app-insights-root-effects';
