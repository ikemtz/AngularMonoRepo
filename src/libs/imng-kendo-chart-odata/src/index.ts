export * from './lib/chart-odata-facade';
export * from './lib/chart-odata-base.component';
export * from './lib/chart-series-data-point';
