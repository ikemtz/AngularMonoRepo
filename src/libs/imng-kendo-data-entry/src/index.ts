export * from './lib/data-entry.module';
export * from './lib/data-entry-dialog.component';
export * from './lib/base-data-entry.component';
export * from './lib/data-entry-facade';
export * from './lib/reducer-collection-logic';
