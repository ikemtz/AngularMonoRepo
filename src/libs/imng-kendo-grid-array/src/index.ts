export * from './lib/imng-kendo-grid-array.module';
export * from './lib/kendo-array-base-component';
export * from './lib/kendo-array-grid.directive';
