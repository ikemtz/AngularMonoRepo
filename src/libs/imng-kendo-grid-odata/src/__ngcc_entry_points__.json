{"ngccVersion":"9.1.0","configFileHash":"87c535c3ce0eac2a54c246892e0e21a1","lockFileHash":"5e332e3bd69b745ec1867eb23dd1d36c","entryPointPaths":[]}