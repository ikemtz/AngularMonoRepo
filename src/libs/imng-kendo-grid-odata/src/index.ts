export * from './lib/kendo-array-component-base';
export * from './lib/kendo-odata-component-base';
export * from './lib/kendo-odata-grid-facade';
export * from './lib/kendo-odata-grid-facade-base';
export * from './lib/kendo-odata-test-facade';
export * from './lib/kendo-odata-selector';
export * from './lib/kendo-odata-grid-state';
export * from './lib/kendo-odata-grid-rxjs-operators';
export * from './lib/kendo-odata-grid-state-change-event';
export * from './lib/kendo-odata-grid-selector-creator';
export * from './lib/kendo-odata-initial-state-creator';
