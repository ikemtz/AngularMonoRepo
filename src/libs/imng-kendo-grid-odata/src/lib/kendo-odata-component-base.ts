import { Observable, isObservable, map } from 'rxjs';
import { PagerSettings } from '@progress/kendo-angular-grid';
import {
  OnInit,
  OnDestroy,
  InjectionToken,
  Inject,
  Component,
} from '@angular/core';
import { ODataState, ODataResult, Expander } from 'imng-kendo-odata';
import { GridStateChangeEvent, KendoGridBaseComponent } from 'imng-kendo-grid';
import { IKendoODataGridFacade } from './kendo-odata-grid-facade';
import { Router } from '@angular/router';
import { Subscribable } from 'imng-ngrx-utils';
import {
  CompositeFilterDescriptor,
  FilterDescriptor,
  isCompositeFilterDescriptor,
} from '@progress/kendo-data-query';

const FACADE = new InjectionToken<IKendoODataGridFacade<unknown>>(
  'imng-grid-odata-facade'
);
const STATE = new InjectionToken<ODataState>('imng-grid-odata-odataState');

@Component({ template: '' })
export abstract class KendoODataBasedComponent<
  ENTITY,
  FACADE extends IKendoODataGridFacade<ENTITY>
  > extends KendoGridBaseComponent<ENTITY>
  implements OnInit, OnDestroy, Subscribable {
  /**
   * This sets the amount of the maximum amount of sortable columns for this component.  Default = 5.
   */
  public maxSortedColumnCount = 5; //NOSONAR
  /**
   * This will allow you to provide a visual indicator that some of the columns have been hidden.
   */
  public hasHiddenColumns$: Observable<boolean>;
  public gridStateQueryKey = 'odataState';
  public gridDataState: ODataState;
  public gridDataResult$: Observable<ODataResult<ENTITY>>;
  public loading$: Observable<boolean>;
  public gridPagerSettings$: Observable<false | PagerSettings>;
  /**
   * A properties enum to make kendo grid columns definitions type safe
   * {@example <kendo-grid-column [field]="props.FIELD_NAME">}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public abstract readonly props: any; //NOSONAR
  protected expanders?: Expander[];
  protected transformations?: string;

  constructor(
    @Inject(FACADE) public readonly facade: FACADE,
    @Inject(STATE) public readonly state: ODataState | Observable<ODataState>,
    public readonly router: Router | null = null, //NOSONAR
    public readonly gridRefresh$: Observable<unknown> | null = null
  ) {
    super();
    if (
      this.router?.routerState?.snapshot?.root.queryParams[
      this.gridStateQueryKey
      ]
    ) {
      try {
        this.gridDataState = this.deserializeODataState(
          this.router?.routerState?.snapshot?.root?.queryParams[
          this.gridStateQueryKey
          ]
        );
      } catch (e) {
        console.error( //NOSONAR
          `Exception thrown while deserializing query string parameter: ${this.gridStateQueryKey}.`
        );
      }
    }
    if (isObservable(state)) {
      this.allSubscriptions.push(
        state.subscribe((t) => {
          this.gridDataState = t;
          this.expanders = t.expanders;
          this.transformations = t.transformations;
        })
      );
    } else {
      this.gridDataState = this.gridDataState
        ? {
          ...this.gridDataState,
          selectors: state.selectors,
          expanders: state.expanders,
        }
        : state;
      this.expanders = state.expanders;
      this.transformations = state.transformations;
    }
    if (gridRefresh$) {
      this.allSubscriptions.push(
        gridRefresh$.subscribe(() => this.loadEntities(this.gridDataState))
      );
    }
  }

  public ngOnInit(): void {
    if (!this.gridRefresh$) {
      this.loadEntities(this.gridDataState);
    }
    this.loading$ = this.facade.loading$;
    this.gridDataResult$ = this.facade.gridData$?.pipe(map(gridData =>
      gridData ? gridData : { total: 0, data: [] }));
    this.gridPagerSettings$ = this.facade.gridPagerSettings$;
  }

  public deserializeODataState(stateQueryParam: string): ODataState {
    const state: ODataState = JSON.parse(atob(stateQueryParam));
    state.filter?.filters?.forEach((filter) => this.normalizeFilters(filter));
    return state;
  }

  public normalizeFilters(
    filter: FilterDescriptor | CompositeFilterDescriptor
  ) {
    if (isCompositeFilterDescriptor(filter)) {
      this.normalizeFilters(filter);
    } else if (
      (filter?.field as string)?.toUpperCase().endsWith('DATE') ||
      (filter?.field as string)?.toUpperCase().endsWith('UTC')
    ) {
      filter.value = new Date(filter.value);
    }
  }

  public serializeODataState(odataState: ODataState): string {
    return btoa(JSON.stringify(odataState));
  }
  /**
   * Will reset filters to initialGrid state passed into the constructor
   */
  public resetFilters(): void {
    if (!isObservable(this.state)) {
      this.gridDataState = {
        ...this.gridDataState,
        filter: this.state.filter,
        inFilters: this.state.inFilters,
        childFilters: this.state.childFilters,
      };
      this.loadEntities(this.gridDataState);
    }
  }

  public dataStateChange(state: GridStateChangeEvent | ODataState): void {
    this.gridDataState = {
      ...state,
      expanders: this.expanders,
      transformations: this.transformations,
    };
    this.loadEntities(this.gridDataState);
  }

  public excelData = (): Observable<ODataResult<ENTITY>> =>
    this.gridDataResult$;

  public loadEntities(odataState: ODataState): void {
    odataState = this.validateSortParameters(odataState);
    this.gridDataState = odataState;
    this.expanders = odataState.expanders;
    this.transformations = odataState.transformations;
    this.facade.loadEntities(this.gridDataState);
    this.updateRouterState(odataState);
  }

  public validateSortParameters(state: ODataState): ODataState {
    if (state.sort && (state.sort?.length || 0) > this.maxSortedColumnCount) {
      state = {
        ...state,
        sort: state.sort.slice(0, this.maxSortedColumnCount),
      };
      console.warn(
        `You have exceeded the limit of ${this.maxSortedColumnCount} sorted columns for the current grid. MAX-Sorted-Column-Count`
      ); //NOSONAR
    }
    return state;
  }

  public updateRouterState(state: ODataState): void {
    if (this.router) {
      const tempState = { ...state };
      delete tempState.selectors;
      delete tempState.expanders;
      this.router.navigate([], {
        relativeTo: this.router.routerState.root,
        queryParams: {
          [this.gridStateQueryKey]: this.serializeODataState(tempState),
        },
        skipLocationChange: false,
        queryParamsHandling: 'merge',
      });
    }
  }
  public reloadEntities(): void {
    this.facade.reloadEntities();
  }
}
