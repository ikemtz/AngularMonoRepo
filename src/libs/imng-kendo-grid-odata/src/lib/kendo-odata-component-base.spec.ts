import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { KendoODataComponentBase } from './kendo-odata-component-base';
import { ODataGridMockFacade, createODataGridMockFacade } from '../../testing/src';
import { readFirst } from '@nrwl/angular/testing/src/testing-utils';
import { FilterDescriptor } from '@progress/kendo-data-query';
import { ODataState } from 'imng-kendo-odata';

describe('KendoODataComponentBase', () => {
  let component: KendoODataGridTestComponent;
  let fixture: ComponentFixture<KendoODataGridTestComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [KendoODataGridTestComponent],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(KendoODataGridTestComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(component.router).toBeFalsy();
  });

  it('should export to Excel', async () => {
    const data = await readFirst(component.excelData());
    expect(data).toStrictEqual({ data: [{ id: 'apples' }], total: 0 });
  });

  it('should reset', async () => {
    component.gridDataState = {
      ...component.gridDataState,
      filter: { logic: 'and', filters: [{ field: 'y', operator: 'contains', value: 56 }] },
    };
    component.resetFilters();
    expect(component.gridDataState).toMatchSnapshot();
  });

  it('should reload', async () => {
    component.reloadEntities();
    expect(component.facade.reloadEntities).toBeCalledTimes(1);
  });

  it('should serialize/deserialize odataState filters correctly', () => {
    const tempDate = new Date();
    const serializedResult = component.serializeODataState({
      filter: { logic: 'and', filters: [{ field: 'xyzDate', operator: 'eq', value: tempDate }] },
    });
    const deserializedResult = component.deserializeODataState(serializedResult);
    expect(tempDate).toEqual((deserializedResult.filter.filters[0] as FilterDescriptor).value);
  });

  it('should limit odataState sort parameters', () => {
    const odataState: ODataState = {
      sort: [
        { field: 'a', dir: 'asc' },
        { field: 'b', dir: 'asc' },
        { field: 'c', dir: 'asc' },
        { field: 'd', dir: 'asc' },
        { field: 'e', dir: 'asc' },
        { field: 'f', dir: 'asc' },
      ],
    };
    component.facade.loadEntities = jest.fn();
    component.loadEntities(odataState);
    expect(component.facade.loadEntities).toBeCalledTimes(1);
    expect(component.facade.loadEntities).toBeCalledWith({
      sort: [
        { field: 'a', dir: 'asc' },
        { field: 'b', dir: 'asc' },
        { field: 'c', dir: 'asc' },
        { field: 'd', dir: 'asc' },
        { field: 'e', dir: 'asc' },
      ],
    });
  });

  it('should serialize/deserialize odataState correctly', () => {
    const serializedResult = component.serializeODataState({});
    const deserializedResult = component.deserializeODataState(serializedResult);
    expect(deserializedResult).toStrictEqual({});
  });
});

const initialGridState: ODataState = {
  selectors: ['x', 'y', 'z'],
  sort: [{ field: 'x', dir: 'desc' }],
  filter: { logic: 'and', filters: [{ field: 'x', operator: 'eq', value: 1 }] },
};
@Component({
  selector: 'imng-test-component',
  template: '<h1></h1>',
})
// eslint-disable-next-line @typescript-eslint/ban-types
export class KendoODataGridTestComponent extends KendoODataComponentBase<object, ODataGridMockFacade> {
  props = {};
  constructor() {
    super(createODataGridMockFacade(), initialGridState);
  }
}
