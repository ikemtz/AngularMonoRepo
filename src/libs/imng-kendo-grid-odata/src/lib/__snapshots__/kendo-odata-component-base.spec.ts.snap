// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KendoODataComponentBase should reset 1`] = `
Object {
  "childFilters": undefined,
  "filter": Object {
    "filters": Array [
      Object {
        "field": "x",
        "operator": "eq",
        "value": 1,
      },
    ],
    "logic": "and",
  },
  "inFilters": undefined,
  "selectors": Array [
    "x",
    "y",
    "z",
  ],
  "sort": Array [
    Object {
      "dir": "desc",
      "field": "x",
    },
  ],
}
`;
